dao-voting/tests/harness.rs

```diff
@@ -0,0 +1,1156 @@
+
+            mint(
+                &deployer.gov_token.as_ref().unwrap(),
+                asset_amount,
+                user.wallet.address(),
+            )
+            .await;
+
+            let call_params = CallParameters::new(
+                Some(asset_amount),
+                Some(AssetId::from(*gov_token_id)),
+                Some(100_000),
+            );
+            deposit(&user.dao_voting, call_params).await;
+
+            let proposal_transaction = proposal_transaction(gov_token_id);
+            create_proposal(&user.dao_voting, 10, 100, proposal_transaction.clone()).await;
+            vote(&user.dao_voting, false, 0, asset_amount / 2).await;
+
+            execute(&user.dao_voting, 0).await;
+        }
+    }
+}
+
+mod unlock_votes {
+    use super::*;
+
+    mod success {
+        use super::*;
+
+        #[tokio::test]
+        async fn user_can_unlock_tokens() {
+            let (_gov_token, gov_token_id, deployer, user, asset_amount) = setup().await;
+            constructor(&deployer.dao_voting, gov_token_id).await;
+
+            mint(
+                &deployer.gov_token.as_ref().unwrap(),
+                asset_amount,
+                user.wallet.address(),
+            )
+            .await;
+
+            let call_params = CallParameters::new(
+                Some(asset_amount),
+                Some(AssetId::from(*gov_token_id)),
+                Some(100_000),
+            );
+            deposit(&user.dao_voting, call_params).await;
+
+            let proposal_transaction = proposal_transaction(gov_token_id);
+            create_proposal(&user.dao_voting, 1, 1, proposal_transaction.clone()).await;
+            vote(&user.dao_voting, true, 0, asset_amount / 2).await;
+
+            assert_eq!(
+                user_balance(&user.dao_voting, Identity::Address(user.wallet.address())).await,
+                asset_amount / 2
+            );
+            assert_eq!(
+                user_votes(
+                    &user.dao_voting,
+                    Identity::Address(user.wallet.address()),
+                    0
+                )
+                .await,
+                Votes {
+                    yes_votes: asset_amount / 2,
+                    no_votes: 0
+                }
+            );
+
+            unlock_votes(&user.dao_voting, 0).await;
+
+            assert_eq!(
+                user_balance(&user.dao_voting, Identity::Address(user.wallet.address())).await,
+                asset_amount
+            );
+            assert_eq!(
+                user_votes(
+                    &user.dao_voting,
+                    Identity::Address(user.wallet.address()),
+                    0
+                )
+                .await,
+                Votes {
+                    yes_votes: 0,
+                    no_votes: 0
+                }
+            );
+        }
+
+        #[tokio::test]
+        async fn user_can_unlock_tokens_from_multiple_proposals() {
```

Comment:
This tests to see if a user can use the same tokens to vote on 1 proposal after another (which is okay) but we also need a test to see if the user can unlock tokens while there are 2 proposals active simultaneously

Reply:
hmm currently a user can only unlock votes if the proposal is expired.  That is all votes are final and cannot be changed.  I admit this may not be the best design decision, but we can handle that in a future pr.

Reply:
What I mean is
- User has 100 tokens
- 2 Proposals exist
- User deposits and votes some % to each
- Both both expire
- Convert into deposit

Rather than
- User has 100 tokens
- Create 1 proposal
- Deposit and vote all
- Expires
- Convert
- Create new proposal
- ...

The point is to make sure that you can withdraw / convert / whatever while more than 1 proposal already exists in state